sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RetryPolicy.java

```diff
@@ -128,46 +135,74 @@ private Mono<HttpResponse> attemptAsync(final HttpPipelineCallContext context, f
                 if (shouldRetry(httpResponse, tryCount)) {
                     final Duration delayDuration = determineDelayDuration(httpResponse, tryCount, retryStrategy,
                         retryAfterHeader, retryAfterTimeUnit);
-                    LOGGER.atVerbose()
-                        .addKeyValue(LoggingKeys.TRY_COUNT_KEY, tryCount)
-                        .addKeyValue(LoggingKeys.DURATION_MS_KEY, delayDuration.toMillis())
-                        .log("Retrying.");
-
-                    Flux<ByteBuffer> responseBody = httpResponse.getBody();
-                    if (responseBody == null) {
-                        return attemptAsync(context, next, originalHttpRequest, tryCount + 1)
-                            .delaySubscription(delayDuration);
-                    } else {
-                        return httpResponse.getBody()
-                            .ignoreElements()
-                            .then(attemptAsync(context, next, originalHttpRequest, tryCount + 1)
-                                .delaySubscription(delayDuration));
-                    }
+                    logRetry(tryCount, delayDuration);
+
+                    httpResponse.close();
+
+                    return attemptAsync(context, next, originalHttpRequest, tryCount + 1)
+                        .delaySubscription(delayDuration);
                 } else {
                     if (tryCount >= retryStrategy.getMaxRetries()) {
-                        LOGGER.atInfo()
-                            .addKeyValue(LoggingKeys.TRY_COUNT_KEY, tryCount)
-                            .log("Retry attempts have been exhausted.");
+                        logRetryExhausted(tryCount);
                     }
                     return Mono.just(httpResponse);
                 }
             })
             .onErrorResume(Exception.class, err -> {
                 if (shouldRetryException(err, tryCount)) {
-                    LOGGER.atVerbose()
-                        .addKeyValue(LoggingKeys.TRY_COUNT_KEY, tryCount)
-                            .log("Error resume.", err);
+                    logRetryWithError(LOGGER.atVerbose(), tryCount, "Error resume.", err);
                     return attemptAsync(context, next, originalHttpRequest, tryCount + 1)
                         .delaySubscription(retryStrategy.calculateRetryDelay(tryCount));
                 } else {
-                    LOGGER.atError()
-                        .addKeyValue(LoggingKeys.TRY_COUNT_KEY, tryCount)
-                        .log("Retry attempts have been exhausted.", err);
+                    logRetryWithError(LOGGER.atError(), tryCount, "Retry attempts have been exhausted.", err);
                     return Mono.error(err);
                 }
             });
     }
 
+    private HttpResponse attemptSync(final HttpPipelineCallContext context, final HttpPipelineNextSyncPolicy next,
```

Comment:
It would have been good to share the code, but the threading/delay abstractions will increase the branching logic and code complexity to know if we want async or sync impl.

I was thinking the accidental skew can be caught in tests?